### PR TITLE
Corrupted model/database crash fix

### LINKED ISSUE #8940: [Bug] Crash 2.15.0 - "flds" is possibly null on Model?
###### Reproduction Steps

1. Load some sort of corrupted collection (I imagine), where there is a note that has a card template / model with no fields
2. Attempt to get a card in the Reviewer that is associated with that note type / broken model

###### Expected Result

A dialog saying your collection is corrupt, and in the database corruption fixes we should repair note types with no fields so that they have valid defaults?

Connected with this - an investigation upstream to see if + and how they handle corrupt models

###### Actual Result

```

Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'com.ichi2.utils.JSONArray com.ichi2.utils.JSONObject.getJSONArray(java.lang.String)' on a null object reference
at com.ichi2.libanki.Models.fieldMap(Models.java:1)
at com.ichi2.libanki.Note.load(Note.java:14)
at com.ichi2.libanki.Note.<init>(Note.java:4)
at com.ichi2.libanki.Collection.getNote(Collection.java:1)
at com.ichi2.libanki.Card.note(Card.java:3)
at com.ichi2.libanki.Card._getQA(Card.java:4)
at com.ichi2.libanki.Card._getQA(Card.java:2)
at com.ichi2.async.CollectionTask$GetCard.task(CollectionTask.java:5)
at com.ichi2.async.CollectionTask$GetCard.task(CollectionTask.java:1)
at com.ichi2.async.CollectionTask.actualDoInBackground(CollectionTask.java:14)
at com.ichi2.async.CollectionTask.doInBackground(CollectionTask.java:2)
at com.ichi2.async.CollectionTask.doInBackground(CollectionTask.java:1)
```

###### Debug info

Seen on ACRA from release tag v.2.15.0

###### Research
*Enter an [x] character to confirm the points below:*

- [ ] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid
- [ ] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [ ] I have searched for similar existing issues here and on the user forum
- [ ] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

### PR DESCRIPTION
**Pull request by dorrin-sot**
## Purpose / Description
App would crash if the database/model was corrupt

## Fixes
Fixes #8940

## Approach
Using the helper methods I added into the `AnkiActivity` class, a corruption check on the database is performed before activity creations and collection loadings. If it goes to deck picker page, shows the dialog and performs a database check and repair if the user selects the option to do so.

## How Has This Been Tested?

https://user-images.githubusercontent.com/59933477/119974166-6b9f5a00-bfc5-11eb-802d-37ee5515c8cf.mov

## Checklist

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
This doesn't seem like an acceptable check to perform on every screen load as it'll be extremely slow.

It'd be better to catch the bubbled exception in the Task, and show a "database corrupt" message, like most of our code does.

Then add a few unit tests for various screens to ensure that they open and work with a corrupt DB

See usages of: https://github.com/ankidroid/Anki-Android/blob/4c30d333d92033c9f48143bdf51074055410abae/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java#L168

**Comment by david-allison-1** [comment]
> @david-allison-1 TBH I did try throwing exceptions to direct the issue to activities to display the dialogs, But got stuck at places where it would lead to "CollectionTask". Is there a way I can display the dialogs from there without adding a weakreference of the activity to its constructor?

Check the references to the variable that I linked, it's typically handled from a listener to the task, possibly converting to a result-like structure or Boolean

**Comment by dorrin-sot** [comment]
Sorry 'bout the atomic commit. New approach is:

1. Everything is as normal until a corruption in the model is noticed (eg. it being null). Then an Exception (`CorruptDatabaseException` that was added for just this purpose) will be thrown.
2. The exception will travel out via `throws` until we get to a method in an activity or task class.
3. if it's an activity class it would catch that exception and call `showDbCorruptDialog`
4. if not it would use a listener class (`ShowDbCorruptDialogListener`) that also calls `showDbCorruptDialog`

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
This change adds a lot of noise to the surrounding code. I don't feel this is necessary, but I'm happy to be mistaken.

The way I see it:

* All significant `libAnki` access should ideally be in `CollectionTasks` (otherwise, we block the UI)
* All CollectionTasks should be able to return a `Pair<Boolean, X>` (a result type) if necessary.
* Most error handling can be done by handling the listeners to the `CollectionTask`
* The main aim would be determining where errors can be raised, returning a result-like structure, and handling this with a "database corrupt".
* If the database is corrupt (a bad model ref for example), then it would be ideal to check for it in the single place where it first crashes, and then assume that it's fine, rather than propagate the error handling throughout the codebase: this adds noise to the code, and logically can't occur if we catch the first occurrence

**Comment by dorrin-sot** [comment]
> This change adds a lot of noise to the surrounding code. I don't feel this is necessary, but I'm happy to be mistaken.
> 
> The way I see it:
> 
> * All significant `libAnki` access should ideally be in `CollectionTasks` (otherwise, we block the UI)
> * All CollectionTasks should be able to return a `Pair<Boolean, X>` (a result type) if necessary.
> * Most error handling can be done by handling the listeners to the `CollectionTask`
> * The main aim would be determining where errors can be raised, returning a result-like structure, and handling this with a "database corrupt".
> * If the database is corrupt (a bad model ref for example), then it would be ideal to check for it in the single place where it first crashes, and then assume that it's fine, rather than propagate the error handling throughout the codebase: this adds noise to the code, and logically can't occur if we catch the first occurrence


I do agree it's really ugly this way. I also have to say another way I tried that would be cleaner (and maybe faster) was in the beginning where I'd call to `DialogHandler` to open the dialog from wherever the error occurred; only problem was It would only handle the message when `onResume` was called so for example only when the back button was pressed. so not only would it not open the dialog on the desired activity, the context in the handler was also too old and couldn't be used for the alert dialog's context. 
Is this way valid? if so, how do I fix this issue?

**Comment by david-allison-1** [comment]
In the code sample linked, we discard the current activity and open the Deck Picker with a code to specify corruption.

There is a rare bug here (cast to DeckPicker failed) that I've been meaning to deal with for some time, but since you're discarding the activity you shouldn't need to worry about additional lifecycle events (other than those on the path to onDestroy

**Comment by dorrin-sot** [comment]
@david-allison-1 sorry, didn't think it would take this long and I've got a bunch of exams coming up :(. Is it ok if I continue working on this in a few weeks?

**Comment by david-allison-1** [comment]
> @david-allison-1 sorry, didn't think it would take this long and I've got a bunch of exams coming up :(. Is it ok if I continue working on this in a few weeks?

Absolutely! Life takes priority. Good luck

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java [hunk 1]
```diff
@@ -547,6 +550,23 @@ protected void showSimpleMessageDialog(String title, @Nullable String message, b
     }
 
 
+    public void showDbCorruptDialog() {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java:553]: We should ideally keep this in the "Main Screen" (currently the deck picker). We already have a dialog for this

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java [hunk 2]
```diff
@@ -547,6 +550,23 @@ protected void showSimpleMessageDialog(String title, @Nullable String message, b
...
+        Timber.d("showDbCorruptDialog()");
+        new AlertDialog.Builder(this)
+                .setTitle("Corrupt Database")
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java:556]: If we are going to use this (I'd expect it to be removed), strings need to be translatable

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java [hunk 3]
```diff
@@ -843,6 +844,9 @@ protected void onActivityResult(int requestCode, int resultCode, Intent intent)
         } else if (resultCode == RESULT_DB_ERROR) {
             handleDbError();
             return;
+        } else if (resultCode == RESULT_MODEL_CORRUPT) {
+            showDbCorruptDialog();
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:848]: We already have a dialog for a DB error. Is there a reason we're not using this?

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/MediaCheckDialog.java [hunk 4]
```diff
@@ -103,17 +103,17 @@ public MaterialDialog onCreateDialog(Bundle savedInstanceState) {
                 // If we have unused files, show a dialog with a "delete" button. Otherwise, the user only
                 // needs to acknowledge the results, so show only an OK dialog.
                 if (unused.size() > 0) {
-                    builder.positiveText(res().getString(R.string.dialog_ok))
-                            .negativeText(res().getString(R.string.check_media_delete_unused))
-                            .onPositive((dialog, which) -> ((MediaCheckDialogListener) getActivity())
+                    builder.negativeText(res().getString(R.string.dialog_ok))
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/MediaCheckDialog.java:106]: What's going on here? Shouldn't OK be positive?

File: AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java [hunk 5]
```diff
@@ -324,7 +325,12 @@ public boolean isFromReviewer() {
             Card newCard = sched.getCard();
             if (newCard != null) {
                 // render cards before locking database
-                newCard._getQA(true);
+                try {
+                    newCard._getQA(true);
+                } catch (CorruptModelException e) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:330]: I feel this should be a more general exception handler

File: AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java [hunk 6]
```diff
@@ -1153,7 +1159,12 @@ public SearchCards(String query, boolean order, int numCardsToRender, int column
                     Timber.d("doInBackgroundSearchCards was cancelled so return null");
                     return null;
                 }
-                searchResult.get(i).load(false, mColumn1Index, mColumn2Index);
+                try {
+                    searchResult.get(i).load(false, mColumn1Index, mColumn2Index);
+                } catch (CorruptModelException e) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:1164]: likewise: probably better as a general exception handler

File: AnkiDroid/src/main/java/com/ichi2/libanki/Models.java [hunk 7]
```diff
@@ -267,7 +267,10 @@ public Model current() {
     public Model current(boolean forDeck) {
         Model m = null;
         if (forDeck) {
-            m = get(mCol.getDecks().current().optLong("mid", -1));
+            try {
+                m = get(mCol.getDecks().current().optLong("mid", -1));
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Models.java:271]: We shouldn't go about changing libAnki code. 

This code needs to be consistent with the Anki Desktop backend, as we're making a transition to this (see ModelsV16.kt)

**Review by david-allison-1 — COMMENTED** [review]
Sorry for the delay - missed the notification. In future, please follow up aggressively. 

This is entirely on me, and by no means an attempt to pass any sort of blame.

**Comment by mikehardy** [comment]
Can always reopen but based on comments and conflict state seems better to start fresh on this one

**Pull request closed without merging**